media-sound/zyn-fusion/zyn-fusion-3.0.5.ebuild

```diff
@@ -0,0 +1,116 @@
+# Copyright 2020 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+
+USE_RUBY="ruby25 ruby26 ruby27"
+PYTHON_COMPAT=( python{2_7,3_6,3_7,3_8,3_9} )
+
+inherit bash-completion-r1 ruby-single python-any-r1
+
+DESCRIPTION="Zyn-Fusion User Interface"
+HOMEPAGE="https://github.com/mruby-zest/mruby-zest-build"
+
+SUBMODULES=(
+	"4f57a1ef9f968e9d5eef53667c7960a2e98c9750 deps/mruby-complex https://github.com/pbosetti/mruby-complex"
```

Comment:
I'm wondering if it wouldn't be better to simply us git-r3 together with the [mruby-zest-duilb repo](https://github.com/mruby-zest/mruby-zest-build/blob/v3.0.2/.gitmodules)?
That seems a lot easier to maintain than this, in case of an update we'd have to update all these submodules which seems pretty annoying tbh.

Reply:
That's not valid for non-live ebuilds; in particular, it's not just the top-level submodules, but also the mruby stuff starts cloning repos down the line if you don't feed them to it ahead of time, during src_compile, and that's a definite no-no. git-r3 fetches in src_unpack which is also no good for regular ebuilds. Non-live ebuilds need to have everything fetched in as a SRC_URI, they can't go out and hit the internet themselves.

The SUBMODULES array is formatted and sorted the same way as `git submodule status`, so at least it should be pretty easy to copy and paste a pile of new hashes for an update.

Reply:
AFAIK git-r3 should work, it checks out submodules recursively. I've used it in the past as well without issues. I'll try it out.

Reply:
Hmm, OK. That introduces a lot of extra work for maintaining packages, similar to the dependency versioning of go and rust packages. Gentoo/the gentoo devs are not being very practical here imho.
Anyway, I guess for now this means this is what it is? We'll just have to maintain a rather large list of dependencies and their versions/git commits?
What could/should we do to improve this? Ask upstream to create release tarballs that include all dependencies?

Reply:
I dunno, I agree with the Gentoo devs that packages should be using proper distfiles-capable tarballs. What they do for packages that don't do this properly is roll their own tarballs. VCS ebuilds are fine for development, but they aren't nice for users... lots of subtle details break when you have an ebuild go out and pull things from the internet. For one, that loses the digest check, so you can't guarantee package authenticity (it being git gives you some of that back, but it's not perfect).

The correct solution is indeed for upstream to release proper tarballs with everything (not the github autoarchives). That's what `www-servers/h2o` has (another package using mruby), which is why it doesn't have this problem.

That said, it isn't really a lot of work. Just checkout the git tag, and paste the output of `git submodule status | cut -d" " -f2-3` in in block overwrite mode, if it's just commit changes for the existing set of deps.